front/components/agent_builder/capabilities/mcp/MCPServerViewsDialog.tsx

```diff
@@ -0,0 +1,694 @@
+  }, [defaultMCPServerViews, nonDefaultMCPServerViews]);
+
+  const toggleToolSelection = (tool: SelectedTool): void => {
+    setSelectedToolsInDialog((prev) => {
+      const isSelected = prev.some((selected) => {
+        if (tool.type !== selected.type) {
+          return false;
+        }
+        if (tool.type === "DATA_VISUALIZATION") {
+          return true;
+        }
+        return (
+          tool.type === "MCP" &&
+          selected.type === "MCP" &&
+          tool.view.sId === selected.view.sId
+        );
+      });
+
+      if (isSelected) {
+        return prev.filter((selected) => {
+          if (tool.type !== selected.type) {
+            return true;
+          }
+          if (tool.type === "DATA_VISUALIZATION") {
+            return false;
+          }
+          return (
+            tool.type === "MCP" &&
+            selected.type === "MCP" &&
+            tool.view.sId !== selected.view.sId
+          );
+        });
+      } else {
+        return [...prev, tool];
+      }
+    });
+  };
+
+  // Data Visualization is not an action but we show like an action in UI.
+  const onClickDataVisualization = () => {
+    if (!dataVisualization) {
+      return;
+    }
+    toggleToolSelection({ type: "DATA_VISUALIZATION" });
+  };
+
+  function onClickMCPServer(mcpServerView: MCPServerViewType) {
+    const tool = { type: "MCP", view: mcpServerView } satisfies SelectedTool;
+    const requirement = getMCPServerRequirements(mcpServerView);
+
+    if (!requirement.noRequirement) {
+      const action = getDefaultMCPAction(mcpServerView);
+      const isReasoning = requirement.requiresReasoningConfiguration;
+
+      if (action.type === "MCP" && isReasoning) {
+        if (reasoningModels.length === 0) {
+          sendNotification({
+            title: "No reasoning model available",
+            description:
+              "Please add a reasoning model to your workspace to be able to use this tool",
+            type: "error",
+          });
+          return;
+        }
+
+        const defaultReasoningModel =
+          reasoningModels.find(
+            (model) => model.modelId === DEFAULT_REASONING_MODEL_ID
+          ) ?? reasoningModels[0];
+
+        setConfigurationTool({
+          ...action,
+          configuration: {
+            ...action.configuration,
+            reasoningModel: {
+              modelId: defaultReasoningModel.modelId,
+              providerId: defaultReasoningModel.providerId,
+              temperature: null,
+              reasoningEffort: null,
+            },
+          },
+        });
+      } else {
+        setConfigurationTool(action);
+      }
+
+      setConfigurationMCPServerView(mcpServerView);
+      setCurrentPageId(CONFIGURATION_DIALOG_PAGE_IDS.CONFIGURATION);
+      return;
+    }
+
+    // No configuration required, add to selected tools
+    toggleToolSelection(tool);
+  }
+
+  const handleAddSelectedTools = useCallback(async () => {
```

Comment:
Except if I missed a piece of code, async is not necessary.
```suggestion
  const handleAddSelectedTools = useCallback(() => {
```